src/py21cmfast/drivers/coeval.py

```diff
@@ -0,0 +1,901 @@
+        """Save to disk.
+
+        This function has defaults that make it easy to save a unique box to
+        the current directory.
+
+        Parameters
+        ----------
+        fname : str, optional
+            The filename to write, default a unique name produced by the inputs.
+        direc : str, optional
+            The directory into which to write the file. Default is the current directory.
+
+        Returns
+        -------
+        str :
+            The filename to which the box was written.
+        """
+        return self._write(direc=direc, fname=fname, clobber=clobber)
+
+    @classmethod
+    def _read_inputs(cls, fname):
+        kwargs = {}
+        with h5py.File(fname, "r") as fl:
+            glbls = dict(fl["_globals"].attrs)
+            kwargs["redshift"] = fl.attrs["redshift"]
+
+            if "photon_nonconservation_data" in fl.keys():
+                d = fl["photon_nonconservation_data"]
+                kwargs["photon_nonconservation_data"] = {k: d[k][...] for k in d.keys()}
+
+        return kwargs, glbls
+
+    @classmethod
+    def read(cls, fname, direc="."):
+        """Read a lightcone file from disk, creating a LightCone object.
+
+        Parameters
+        ----------
+        fname : str
+            The filename path. Can be absolute or relative.
+        direc : str
+            If fname, is relative, the directory in which to find the file. By default,
+            both the current directory and default cache and the  will be searched, in
+            that order.
+
+        Returns
+        -------
+        LightCone :
+            A :class:`LightCone` instance created from the file's data.
+        """
+        if not os.path.isabs(fname):
+            fname = os.path.abspath(os.path.join(direc, fname))
+
+        if not os.path.exists(fname):
+            raise FileExistsError(f"The file {fname} does not exist!")
+
+        park, glbls = cls._read_inputs(fname)
+        boxk = cls._read_particular(fname)
+
+        with global_params.use(**glbls):
+            out = cls(**park, **boxk)
+
+        return out
+
+
+class Coeval(_HighLevelOutput):
+    """A full coeval box with all associated data."""
+
+    def __init__(
+        self,
+        redshift: float,
+        initial_conditions: InitialConditions,
+        perturbed_field: PerturbedField,
+        ionized_box: IonizedBox,
+        brightness_temp: BrightnessTemp,
+        ts_box: TsBox | None = None,
+        halobox: HaloBox | None = None,
+        cache_files: dict | None = None,
+        photon_nonconservation_data=None,
+        _globals=None,
+    ):
+
+        # Check that all the fields have the same input parameters.
+        InputParameters.from_output_structs(
+            (
+                initial_conditions,
+                perturbed_field,
+                halobox,
+                ionized_box,
+                brightness_temp,
+                ts_box,
+            )
+        )
```

Comment:
Should these be saved to the struct itself instead of taking the brightness_temp struct's parameters?

Reply:
We certainly _could_ do that -- at the moment we always require having a `brightness_temp` struct in this instance, and the various input structs _must_ be the same for all boxes, so merely referencing them from the `brightness_temp` struct makes sense to me